src/Http/Routing/src/EndpointSelectorContext.cs

```diff
@@ -24,7 +24,7 @@ public sealed class EndpointSelectorContext : IEndpointFeature, IRouteValuesFeat
         /// </summary>
         public RouteValueDictionary RouteValues
         {
-            get => _routeValues;
+            get => _routeValues ?? (_routeValues = new RouteValueDictionary());
```

Comment:
If we're going to do this, we can take advantage of it in more cases, including the ones that involve cnadidate set.

Reply:
This is beyond the scope of this PR, but what about reusing RVD instances? It would be similar to how we're reusing HttpContext et al. It's basically a per-request allocations that is tied to a feature.

Reusing RVD, and the routing feature, would save allocations on every request, not just those with no route values.

[Caching per-request objects on reused HttpContexts](https://github.com/aspnet/AspNetCore/issues/6895)

Reply:
We can just straight up do this in Kestrel I fear we won't be able to do the more generic caching for 3.0

Reply:
> If we're going to do this, we can take advantage of it in more cases, including the ones that involve cnadidate set.

Yes, we can, I just wanted to minimize the code change for now. I have to look at see what things are assuming route values are never null when looking at CandidateState.RouteValues (maybe I can do it statiaclly with #enable nullable 😄 )

Reply:
CandidateSet has some special semantics around RVD that should just be removed. If treats null RVD as a shorthand for not-a-match in some APIs.